models/repo/setting.go

```diff
@@ -0,0 +1,148 @@
+	for _, s := range settings {
+		settingsMap[s.SettingKey] = s
+	}
+	return settingsMap, nil
+}
+
+func validateRepoSettingKey(key string) error {
+	if len(key) == 0 {
+		return fmt.Errorf("setting key must be set")
+	}
+	if strings.ToLower(key) != key {
+		return fmt.Errorf("setting key should be lowercase")
+	}
+	return nil
+}
+
+// GetRepoSetting gets a specific setting for a repo
+func GetRepoSetting(repoID int64, key string, def ...string) (string, error) {
+	if err := validateRepoSettingKey(key); err != nil {
+		return "", err
+	}
+	setting := &Setting{RepoID: repoID, SettingKey: key}
+	has, err := db.GetEngine(db.DefaultContext).Get(setting)
+	if err != nil {
+		return "", err
+	}
+	if !has {
+		if len(def) == 1 {
+			return def[0], nil
+		}
+		return "", nil
+	}
+	return setting.SettingValue, nil
+}
+
+// DeleteRepoSetting deletes a specific setting for a repo
+func DeleteRepoSetting(repoID int64, key string) error {
+	if err := validateRepoSettingKey(key); err != nil {
+		return err
+	}
+	_, err := db.GetEngine(db.DefaultContext).Delete(&Setting{RepoID: repoID, SettingKey: key})
+	return err
+}
+
+// SetRepoSetting updates a repos' setting for a specific key
+func SetRepoSetting(repoID int64, key, value string) error {
+	if err := validateRepoSettingKey(key); err != nil {
+		return err
+	}
+	return upsertRepoSettingValue(repoID, key, value)
+}
+
+func upsertRepoSettingValue(repoID int64, key, value string) error {
+	return db.WithTx(func(ctx context.Context) error {
+		e := db.GetEngine(ctx)
+
+		// here we use a general method to do a safe upsert for different databases (and most transaction levels)
+		// 1. try to UPDATE the record and acquire the transaction write lock
+		//    if UPDATE returns non-zero rows are changed, OK, the setting is saved correctly
+		//    if UPDATE returns "0 rows changed", two possibilities: (a) record doesn't exist  (b) value is not changed
+		// 2. do a SELECT to check if the row exists or not (we already have the transaction lock)
+		// 3. if the row doesn't exist, do an INSERT (we are still protected by the transaction lock, so it's safe)
+		//
+		// to optimize the SELECT in step 2, we can use an extra column like `revision=revision+1`
+		//    to make sure the UPDATE always returns a non-zero value for existing (unchanged) records.
+
+		res, err := e.Exec("UPDATE repo_setting SET setting_value=? WHERE setting_key=? AND repo_id=?", value, key, repoID)
+		if err != nil {
+			return err
+		}
+		rows, _ := res.RowsAffected()
+		if rows > 0 {
+			// the existing row is updated, so we can return
+			return nil
+		}
+
+		// in case the value isn't changed, update would return 0 rows changed, so we need this check
+		has, err := e.Exist(&Setting{RepoID: repoID, SettingKey: key})
+		if err != nil {
+			return err
+		}
+		if has {
+			return nil
+		}
+
+		// if no existing row, insert a new row
+		_, err = e.Insert(&Setting{RepoID: repoID, SettingKey: key, SettingValue: value})
+		return err
+	})
```

Comment:
I understand correctly that all of this extra work is done to ensure one atomic operation and that there really only exists one entry in the database, correct?
I think this should be unified into one function, `Upsert` that either gets added to `xorm` directly (best-case),
or as a workaround to `models/db`.
I prefer adding it to xorm directly as
1. xorm is already operating close to the database
2. it is a common request to only have one instance that matches some criteria
and
3. it is less error-prone if it is defined at one place and works for all supported databases instead of being redefined in every package or even in every function.

Reply:
Maybe `db.Upsert(cond, up, ins)`?

Reply:
Why `cond`?